xtask/codegen/gritql.ungram

```diff
@@ -0,0 +1,554 @@
+
+GritCurlyPredicateList =
+    '{'
+    predicates: GritPredicateList?
+    '}'
+
+GritPredicateDefinition =
+    'predicate'
+    name: GritName
+    '('
+    args: GritPatternArgList?
+    ')'
+    body: GritCurlyPredicateList
+
+GritFunctionDefinition =
+    'function'
+    name: GritName
+    '('
+    args: GritVariableList?
+    ')'
+    body: GritCurlyPredicateList
+
+AnyGritPredicate =
+    GritPredicateNot
+    | GritPredicateMaybe
+    | GritPredicateAnd
+    | GritPredicateOr
+    | GritPredicateAny
+    | GritPredicateIfElse
+    | GritPredicateAssignment
+    | GritPredicateAccumulate
+    | GritPredicateRewrite
+    | GritPredicateGreater
+    | GritPredicateLess
+    | GritPredicateGreaterEqual
+    | GritPredicateLessEqual
+    | GritPredicateNotEqual
+    | GritPredicateEqual
+    | GritPredicateMatch
+    | GritPredicateCall
+    | BracketedGritPredicate
+    | GritBooleanLiteral
+    | GritPredicateReturn
+    | GritBogusPredicate
+
+BracketedGritPredicate = '(' AnyGritPredicate ')'
+
+GritPredicateNot =
+    GritNot
+    predicate: AnyGritPredicate
+
+GritPredicateMaybe =
+    'maybe'
+    predicate: AnyGritPredicate
+
+GritPredicateAnd =
+    'and'?
+    '{'
+    predicates: GritPredicateList?
+    '}'
+
+GritPredicateOr =
+    'or'
+    '{'
+    predicates: GritPredicateList?
+    '}'
+
+GritPredicateAny =
+    'any'
+    '{'
+    predicates: GritPredicateList?
+    '}'
+
+GritPredicateIfElse =
+    'if'
+    '('
+    if_predicate: AnyGritPredicate
+    ')'
+    then_predicate: AnyGritPredicate
+    GritPredicateElseClause?
+
+GritPredicateElseClause =
+    'else'
+    else_predicate: AnyGritPredicate
+
+GritPredicateRewrite =
+    left: GritVariable
+    annotation: GritAnnotation?
+    '=>'
+    right: AnyGritPattern
+    GritBogus?
+
+GritPredicateAssignment =
+    container: AnyGritContainer
+    '='
+    pattern: AnyGritPattern
+    GritBogus?
+
+GritPredicateAccumulate =
+    left: GritVariable
+    '+='
+    right: AnyGritPattern
+
+GritPredicateGreater =
+    left: GritVariable
+    '>'
+    right: AnyGritPattern
+
+GritPredicateLess =
+    left: GritVariable
+    '<'
+    right: AnyGritPattern
+
+GritPredicateGreaterEqual =
+    left: GritVariable
+    '>='
+    right: AnyGritPattern
+
+GritPredicateLessEqual =
+    left: GritVariable
+    '<='
+    right: AnyGritPattern
+
+GritPredicateNotEqual =
+    left: GritVariable
+    '!='
+    right: AnyGritPattern
+
+GritPredicateEqual =
+    left: GritVariable
+    '=='
+    right: AnyGritPattern
+
+GritPredicateMatch =
+    left: GritPredicateMatchSubject
+    match: '<:'
+    right: AnyGritPattern
+    GritBogus?
+
+GritPredicateMatchSubject = AnyGritContainer | AnyGritLiteral
+
+GritPredicateCall =
+    name: GritName
+    '('
+    named_args: GritNamedArgList?
+    ')'
+
+GritPredicateReturn =
+    'return'
+    pattern: AnyGritPattern
+
+// --- tokens and lexical definitions
+
+GritBooleanLiteral =
+    value: ('true' | 'false')
+
+GritVariable = 'grit_variable'
```

Comment:
If you intend to have these kind of patterns, you follow the approach we're using the in the other languages. E.g.:

```
JsNullLiteral = value: 'null'
```

Always define a `value:` slot, so the syntax can provide an API to access to the actual variable.